File: server/src/collectors/jmaTide.js
/**
 * JMA Tide Observation Collector
 *
 * Real upstream (scraped from the bosai/tidelevel web app):
 *   - Stations:    https://www.jma.go.jp/bosai/tidelevel/const/tide_area.json
 *                  Maps class20 area code → list of class30 ports → list of
 *                  stations with name, lat, lon, max-recorded level, etc.
 *   - Estimates:   https://www.jma.go.jp/bosai/tidelevel/data/tide/tide_estimate.json
 *                  Keyed on `<class20Code>_<stationCode>` → { estimate, astro }
 *                  arrays (cm). `estimate` is recent observed levels; `astro`
 *                  is the astronomical-tide prediction series.
 *
 * Falls back to a curated seed of major tide gauges if the JMA endpoints
 * fail.
 */

const TIDE_AREA_URL = 'https://www.jma.go.jp/bosai/tidelevel/const/tide_area.json';
const TIDE_ESTIMATE_URL = 'https://www.jma.go.jp/bosai/tidelevel/data/tide/tide_estimate.json';

const SEED_TIDE_STATIONS = [
  { name: '稚内', lat: 45.4082, lon: 141.6864, level_cm: 105, anomaly_cm: 5, region: 'Hokkaido' },
  { name: '函館', lat: 41.7775, lon: 140.7286, level_cm: 95, anomaly_cm: 4, region: 'Hokkaido' },
  { name: '宮古', lat: 39.6447, lon: 141.9711, level_cm: 110, anomaly_cm: 8, region: 'Tohoku' },
  { name: '銚子', lat: 35.7406, lon: 140.8689, level_cm: 92, anomaly_cm: 3, region: 'Kanto' },
  { name: '東京晴海', lat: 35.6500, lon: 139.7700, level_cm: 105, anomaly_cm: 5, region: 'Kanto' },
  { name: '名古屋', lat: 35.0917, lon: 136.8806, level_cm: 110, anomaly_cm: 8, region: 'Tokai' },
  { name: '潮岬', lat: 33.4500, lon: 135.7600, level_cm: 92, anomaly_cm: 3, region: 'Kansai' },
  { name: '神戸', lat: 34.6833, lon: 135.1833, level_cm: 105, anomaly_cm: 6, region: 'Kansai' },
  { name: '高知', lat: 33.5089, lon: 133.5694, level_cm: 100, anomaly_cm: 5, region: 'Shikoku' },
  { name: '長崎', lat: 32.7497, lon: 129.8775, level_cm: 110, anomaly_cm: 6, region: 'Kyushu' },
  { name: '那覇', lat: 26.2125, lon: 127.6809, level_cm: 95, anomaly_cm: 3, region: 'Okinawa' },
];

async function fetchJsonShort(url, timeoutMs = 10000) {
  try {
    const ctrl = new AbortController();
    const t = setTimeout(() => ctrl.abort(), timeoutMs);
    const res = await fetch(url, { signal: ctrl.signal });
    clearTimeout(t);
    if (!res.ok) return null;
    return await res.json();
  } catch { return null; }
}

// Pull the latest non-null observed level (cm) and matching astronomical
// prediction so we can report a meaningful anomaly. JMA's `estimate` can
// have leading null entries for hours not yet reported.
function latestObserved(estimate, astro) {
  if (!Array.isArray(estimate)) return null;
  for (let i = estimate.length - 1; i >= 0; i--) {
    const v = estimate[i];
    if (Number.isFinite(v)) {
      // Astro array spans the day; pick the same hour-of-day index.
      // estimate[] holds the most-recent N hours; the array length tells
      // us how many hours back we are. We map index i to the matching
      // astro slot using the offset from the end of estimate.
      const astroIdx = Array.isArray(astro)
        ? astro.length - (estimate.length - 1 - i) - 1
        : -1;
      const astroVal = astroIdx >= 0 && astroIdx < (astro?.length || 0)
        ? astro[astroIdx]
        : null;
      return {
        level_cm: v,
        astro_cm: Number.isFinite(astroVal) ? astroVal : null,
        hours_ago: estimate.length - 1 - i,
      };
    }
  }
  return null;
}

async function tryJmaTide() {
  const [areas, estimates] = await Promise.all([
    fetchJsonShort(TIDE_AREA_URL),
    fetchJsonShort(TIDE_ESTIMATE_URL),
  ]);
  if (!areas || !estimates || !estimates.areas) return null;

  const features = [];
  // tide_area.json is keyed by class20 area code → { name, class30s: [{ stations:[...] }, ...] }
  for (const [class20Code, area] of Object.entries(areas)) {
    if (!area?.class30s) continue;
    for (const class30 of area.class30s) {
      // The tide_estimate.json keys are `<class20Code>_<class30Code>` —
      // one estimate stream per class30 group, not per individual station.
      const estKey = `${class20Code}_${class30.code}`;
      const est = estimates.areas[estKey];
      const obs = est ? latestObserved(est.estimate, est.astro) : null;

      for (const station of class30?.stations || []) {
        const lat = Number(station.lat);
        const lon = Number(station.lon);
        const geocoded = Number.isFinite(lat) && Number.isFinite(lon);

        features.push({
          type: 'Feature',
          geometry: geocoded ? { type: 'Point', coordinates: [lon, lat] } : null,
          properties: {
            station_id: `JMA_TIDE_${station.code}`,
            name: station.name || `Station ${station.code}`,
            level_cm: obs ? obs.level_cm : null,
            astro_cm: obs ? obs.astro_cm : null,
            anomaly_cm: obs && obs.astro_cm != null ? obs.level_cm - obs.astro_cm : null,
            hours_ago: obs ? obs.hours_ago : null,
            class20_name: area.name || null,
            address: station.addr || null,
            reference: station.reference || null,
            operator_type: station.typeName || station.type || null,
            country: 'JP',
            observed_at: estimates.time || null,
            source: 'jma_bosai_tidelevel',
          },
        });
      }
    }
  }
  return features.length > 0 ? features : null;
}

function generateSeedData() {
  const now = new Date();
  return SEED_TIDE_STATIONS.map((s, i) => ({
    type: 'Feature',
    geometry: { type: 'Point', coordinates: [s.lon, s.lat] },
    properties: {
      station_id: `TIDE_${String(i + 1).padStart(5, '0')}`,
      name: s.name,
      level_cm: s.level_cm,
      anomaly_cm: s.anomaly_cm,
      region: s.region,
      country: 'JP',
      observed_at: now.toISOString(),
      source: 'jma_tide_seed',
    },
  }));
}

export default async function collectJmaTide() {
  let features = await tryJmaTide();
  const live = !!(features && features.length > 0);
  if (!live) features = generateSeedData();
  return {
    type: 'FeatureCollection',
    features,
    _meta: {
      source: 'jma_tide',
      fetchedAt: new Date().toISOString(),
      recordCount: features.length,
      live,
      live_source: live ? 'jma_bosai_tidelevel' : 'jma_tide_seed',
      description: 'JMA tide gauge stations + latest observed level vs astronomical prediction',
    },
  };
}

File: native/collectors/environment/sources/jma_tide.c
/* collectors/environment/sources/jma_tide.c
 * Port of server/src/collectors/jmaTide.js (tryJmaTide).
 * Two JMA bosai JSON fetches: tide_area.json (class20 → class30 → stations)
 * and tide_estimate.json (areas keyed by `<class20>_<class30>`). Builds one
 * Feature per station with latest observed level vs astronomical prediction.
 * SEED fallback intentionally not ported (JS uses it only when live empty).
 */
#include "../../../source.h"
#include "../../../lib/geojson.h"
#include "../../../lib/feedlib.h"
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#define TIDE_AREA_URL     "https://www.jma.go.jp/bosai/tidelevel/const/tide_area.json"
#define TIDE_ESTIMATE_URL "https://www.jma.go.jp/bosai/tidelevel/data/tide/tide_estimate.json"

/* latestObserved: walk estimate[] backwards for a finite value, map to the
 * astro slot by offset from the end. Returns 1 and fills out params if found. */
static int latest_observed(const cJSON *estimate, const cJSON *astro,
                           double *level, int *has_astro, double *astro_v,
                           int *hours_ago) {
  if (!cJSON_IsArray(estimate)) return 0;
  int elen = cJSON_GetArraySize(estimate);
  for (int i = elen - 1; i >= 0; i--) {
    cJSON *v = cJSON_GetArrayItem((cJSON *)estimate, i);
    if (v && cJSON_IsNumber(v) && isfinite(v->valuedouble)) {
      *level = v->valuedouble;
      *hours_ago = elen - 1 - i;
      *has_astro = 0;
      *astro_v = 0;
      if (cJSON_IsArray(astro)) {
        int alen = cJSON_GetArraySize(astro);
        int aidx = alen - (elen - 1 - i) - 1;
        if (aidx >= 0 && aidx < alen) {
          cJSON *av = cJSON_GetArrayItem((cJSON *)astro, aidx);
          if (av && cJSON_IsNumber(av) && isfinite(av->valuedouble)) {
            *has_astro = 1;
            *astro_v = av->valuedouble;
          }
        }
      }
      return 1;
    }
  }
  return 0;
}

/* JS: x || null  — string value or JSON null. */
static cJSON *str_or_null(const cJSON *o, const char *k) {
  cJSON *v = cJSON_GetObjectItem(o, k);
  if (v && cJSON_IsString(v) && v->valuestring[0])
    return cJSON_CreateString(v->valuestring);
  return cJSON_CreateNull();
}

static int run(const source_ctx *ctx, intel_sink *sink) {
  cJSON *areas     = feed_get_json(ctx->http, TIDE_AREA_URL, 10000);
  cJSON *estimates = feed_get_json(ctx->http, TIDE_ESTIMATE_URL, 10000);
  cJSON *estAreas  = estimates ? cJSON_GetObjectItem(estimates, "areas") : NULL;
  if (!areas || !estimates || !estAreas) {
    if (areas) cJSON_Delete(areas);
    if (estimates) cJSON_Delete(estimates);
    return -1;
  }
  cJSON *estTime = cJSON_GetObjectItem(estimates, "time");

  cJSON *features = cJSON_CreateArray();
  cJSON *area;
  cJSON_ArrayForEach(area, areas) {
    const char *class20Code = area->string;
    cJSON *class30s = cJSON_GetObjectItem(area, "class30s");
    if (!cJSON_IsArray(class30s)) continue;
    cJSON *areaName = cJSON_GetObjectItem(area, "name");

    cJSON *class30;
    cJSON_ArrayForEach(class30, class30s) {
      cJSON *c30code = cJSON_GetObjectItem(class30, "code");
      char estKey[128];
      snprintf(estKey, sizeof estKey, "%s_%s",
               class20Code ? class20Code : "",
               (c30code && cJSON_IsString(c30code)) ? c30code->valuestring
                 : (c30code && cJSON_IsNumber(c30code) ? "" : ""));
      if (c30code && cJSON_IsNumber(c30code))
        snprintf(estKey, sizeof estKey, "%s_%g",
                 class20Code ? class20Code : "", c30code->valuedouble);

      cJSON *est = cJSON_GetObjectItem(estAreas, estKey);
      int haveObs = 0, hasAstro = 0, hoursAgo = 0;
      double levelCm = 0, astroCm = 0;
      if (est) {
        cJSON *e = cJSON_GetObjectItem(est, "estimate");
        cJSON *a = cJSON_GetObjectItem(est, "astro");
        haveObs = latest_observed(e, a, &levelCm, &hasAstro, &astroCm,
                                  &hoursAgo);
      }

      cJSON *stations = cJSON_GetObjectItem(class30, "stations");
      if (!cJSON_IsArray(stations)) continue;
      cJSON *station;
      cJSON_ArrayForEach(station, stations) {
        cJSON *latv = cJSON_GetObjectItem(station, "lat");
        cJSON *lonv = cJSON_GetObjectItem(station, "lon");
        double lat = latv ? (cJSON_IsString(latv) ? atof(latv->valuestring)
                              : latv->valuedouble) : NAN;
        double lon = lonv ? (cJSON_IsString(lonv) ? atof(lonv->valuestring)
                              : lonv->valuedouble) : NAN;
        int geocoded = isfinite(lat) && isfinite(lon)
                       && (latv != NULL) && (lonv != NULL);

        cJSON *feat = cJSON_CreateObject();
        cJSON_AddStringToObject(feat, "type", "Feature");
        if (geocoded) {
          cJSON *g = cJSON_CreateObject();
          cJSON_AddStringToObject(g, "type", "Point");
          cJSON *co = cJSON_CreateArray();
          cJSON_AddItemToArray(co, cJSON_CreateNumber(lon));
          cJSON_AddItemToArray(co, cJSON_CreateNumber(lat));
          cJSON_AddItemToObject(g, "coordinates", co);
          cJSON_AddItemToObject(feat, "geometry", g);
        } else {
          cJSON_AddItemToObject(feat, "geometry", cJSON_CreateNull());
        }

        cJSON *p = cJSON_CreateObject();         /* EXACT JS key order */
        cJSON *codev = cJSON_GetObjectItem(station, "code");
        char sidb[96];
        if (codev && cJSON_IsString(codev))
          snprintf(sidb, sizeof sidb, "JMA_TIDE_%s", codev->valuestring);
        else if (codev && cJSON_IsNumber(codev))
          snprintf(sidb, sizeof sidb, "JMA_TIDE_%g", codev->valuedouble);
        else
          snprintf(sidb, sizeof sidb, "JMA_TIDE_");
        cJSON_AddStringToObject(p, "station_id", sidb);
        /* name: station.name || `Station ${station.code}` */
        cJSON *nmv = cJSON_GetObjectItem(station, "name");
        if (nmv && cJSON_IsString(nmv) && nmv->valuestring[0]) {
          cJSON_AddStringToObject(p, "name", nmv->valuestring);
        } else {
          char nb[96];
          if (codev && cJSON_IsString(codev))
            snprintf(nb, sizeof nb, "Station %s", codev->valuestring);
          else if (codev && cJSON_IsNumber(codev))
            snprintf(nb, sizeof nb, "Station %g", codev->valuedouble);
          else
            snprintf(nb, sizeof nb, "Station ");
          cJSON_AddStringToObject(p, "name", nb);
        }
        cJSON_AddItemToObject(p, "level_cm",
          haveObs ? cJSON_CreateNumber(levelCm) : cJSON_CreateNull());
        cJSON_AddItemToObject(p, "astro_cm",
          (haveObs && hasAstro) ? cJSON_CreateNumber(astroCm)
                                : cJSON_CreateNull());
        cJSON_AddItemToObject(p, "anomaly_cm",
          (haveObs && hasAstro) ? cJSON_CreateNumber(levelCm - astroCm)
                                : cJSON_CreateNull());
        cJSON_AddItemToObject(p, "hours_ago",
          haveObs ? cJSON_CreateNumber(hoursAgo) : cJSON_CreateNull());
        cJSON_AddItemToObject(p, "class20_name",
          (areaName && cJSON_IsString(areaName) && areaName->valuestring[0])
            ? cJSON_CreateString(areaName->valuestring) : cJSON_CreateNull());
        cJSON_AddItemToObject(p, "address", str_or_null(station, "addr"));
        cJSON_AddItemToObject(p, "reference",
          str_or_null(station, "reference"));
        /* operator_type: station.typeName || station.type || null */
        cJSON *tn = cJSON_GetObjectItem(station, "typeName");
        cJSON *ty = cJSON_GetObjectItem(station, "type");
        if (tn && cJSON_IsString(tn) && tn->valuestring[0])
          cJSON_AddStringToObject(p, "operator_type", tn->valuestring);
        else if (ty && cJSON_IsString(ty) && ty->valuestring[0])
          cJSON_AddStringToObject(p, "operator_type", ty->valuestring);
        else if (ty && cJSON_IsNumber(ty))
          cJSON_AddItemToObject(p, "operator_type",
            cJSON_Duplicate(ty, 1));
        else
          cJSON_AddItemToObject(p, "operator_type", cJSON_CreateNull());
        cJSON_AddStringToObject(p, "country", "JP");
        cJSON_AddItemToObject(p, "observed_at",
          (estTime && cJSON_IsString(estTime) && estTime->valuestring[0])
            ? cJSON_CreateString(estTime->valuestring) : cJSON_CreateNull());
        cJSON_AddStringToObject(p, "source", "jma_bosai_tidelevel");
        cJSON_AddItemToObject(feat, "properties", p);
        cJSON_AddItemToArray(features, feat);
      }
    }
  }

  cJSON_Delete(areas);
  cJSON_Delete(estimates);

  int n = geojson_emit_features(sink, ctx->source_id, features);
  cJSON_Delete(features);
  fprintf(stderr, "[jma-tide] emitted %d\n", n);
  return n >= 0 ? 0 : -1;
}

static const source_def jma_tide_def = {
  .id = "jma-tide", .collector = "environment",
  .name = "JMA Tide Levels", .name_ja = "気象庁 潮位情報",
   .update_interval_sec = 600, .run = run,
};
REGISTER_SOURCE(jma_tide_def)
